### PR TITLE
Feat(TFD-175): Add Window processing-component

### PR DESCRIPTION
**Pull request by StefanLMNG**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

First implementation of the TCOMP Window component.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
Regarding which processing-component's PR is merged first, the ProcessingFamilyDefinition will be a bit changed to add the other definition components.

### CODE REVIEW

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowDefinition.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    public String[] getFamilies() {
+        return new String[] { ProcessingFamilyDefinition.NAME };
+    }
+
+    public Property[] getReturnProperties() {
+        return new Property[] {};
+    }
+
+    @Override
+    public String getPngImagePath(ComponentImageType imageType) {
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowDefinition.java]: This method is doing exactly what the parent class is doing. Remove it.

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowDefinition.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        switch (imageType) {
+        case PALLETE_ICON_32X32:
+            return "Window_icon32.png"; //$NON-NLS-1$
+        default:
+            return "Window_icon32.png"; //$NON-NLS-1$
+        }
+    }
+
+    @Override
+    public RuntimeInfo getRuntimeInfo(ExecutionEngine engine, ComponentProperties properties,
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowDefinition.java]: Please add all to `assertEngineCompatibility` and `assertConnectorTopologyCompatibility`

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowProperties.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        super.setupProperties();
+        windowLength.setValue(-1);
+        windowSlideLength.setValue(-1);
+        windowSession.setValue(false);
+    }
+
+    @Override
+    public void refreshLayout(Form form) {
+        super.refreshLayout(form);
+        // everything is always visible
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/window/WindowProperties.java]: No, it's not.

File: components/components-processing/processing-runtime/src/main/java/org/talend/components/processing/runtime/window/windowRuntime.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        implements RuntimableRuntime, BeamJobBuilder {
+
+    transient private WindowProperties properties;
+
+    @Override
+    public PCollection<IndexedRecord> expand(PCollection<IndexedRecord> indexedRecordPCollection) {
+        PCollection<IndexedRecord> windowed_items;
+
+        if (properties.windowLength.getValue() < 1) {
+            throw new TalendRuntimeException(CommonErrorCodes.UNEXPECTED_ARGUMENT);
```
Comment (pbailly) [inline components/components-processing/processing-runtime/src/main/java/org/talend/components/processing/runtime/window/windowRuntime.java]: This error code does not explain at all what is the problem.
You should add a cause.

Comment (StefanLMNG) [inline components/components-processing/processing-runtime/src/main/java/org/talend/components/processing/runtime/window/windowRuntime.java]: For the moment I will keep that like that but it should be changed later.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/544/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/544/)

**Pull request merged**